### PR TITLE
@W-15205338 adding new rule for unguarded async ops and eventing

### PR DESCRIPTION
**Pull request by abhagta-sfdc**

### CODE REVIEW

File: test/lib/rules/no-unguarded-async-and-event-operations.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            code: `setTimeout(() => { console.log('Hello'); }, 1000);`, // This should still be valid
+        },
+        {
+            code: `if (!import.meta.env.SSR) { new Promise((resolve) => resolve('test')); }`,
+        },
+        {
+            code: `if (!import.meta.env.SSR) { somePromise.then(() => {}).catch(() => {}); }`,
+        },
+        {
+            code: `if (!import.meta.env.SSR) { this.dispatchEvent(new CustomEvent('myevent')); }`,
```
Comment (lpomerleau) [inline test/lib/rules/no-unguarded-async-and-event-operations.js:35]: The README uses `if (typeof window !== 'undefined')`. Is there a test for that style guard?

File: lib/rule-helpers.js [hunk 2]
```diff
@@ -384,3 +384,90 @@ module.exports.noNodeEnvInSSR = function noNodeEnvInSSR(context) {
...
+
+module.exports.noUnguardedAsyncSSR = function noUnguardedAsyncSSR(context) {
+    const {
+        withinLWCVisitors,
+        isInsideReachableMethod,
+        isInsideReachableFunction,
+        isInsideSkippedBlock,
+    } = reachableDuringSSRPartial();
+
+    const asyncOperations = new Set(['fetch', 'then', 'catch', 'finally']);
```
Comment (lpomerleau) [inline lib/rule-helpers.js:396]: `try` blocks also use `catch` and `finally`. Will this rule get triggered for this?

```
try {
    // do stuff
} catch(e) {
    // handle error
} finally {
    // cleanup
}

File: test/lib/rules/no-unguarded-async-and-event-operations.js (hunk 1 above)
Comment (abhagta-sfdc) [reply test/lib/rules/no-unguarded-async-and-event-operations.js:35]: Thats a miss from my side will update that

File: lib/rule-helpers.js (hunk 2 above)
Comment (abhagta-sfdc) [reply lib/rule-helpers.js:396]: No it wont flag then I will also add a test case for this.

File: docs/rules/no-unguarded-async-and-event-operations.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+### Rule Options
+
+The `no-unguarded-async-and-event` rule can be customized using options to specify additional async operations and event operations that should be monitored for unguarded usage.
+
+#### Options Structure
+
+The rule accepts a single object as its first argument, which can contain the following properties:
+
+-   `additional-async-operations`: An array of strings representing additional async operations that should be checked for \* \* unguarded usage. If you want to include custom async methods that are not part of the default set (e.g., methods specific to your application), you can specify them here.
```
[the remark this answers is not in the document]
Comment (abhagta-sfdc) [reply docs/rules/no-unguarded-async-and-event-operations.md:55]: Fixed

File: docs/rules/no-unguarded-async-and-event-operations.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+-   `additional-event-operations`: An array of strings representing additional event operations to be checked for unguarded usage. Similar to the async operations, this allows you to include custom event handling methods specific to your application.
+
+### When Not To Use This Rule
+
+If your project is strictly running in a browser environment and does not use SSR, you may choose to disable this rule.
```
[the remark this answers is not in the document]
Comment (Shinoni) [reply docs/rules/no-unguarded-async-and-event-operations.md]: @lpomerleau , should we wait till we merge all new rules, then cut a new release and add them to the SSR preset?

[the remark this answers is not in the document]
Comment (lpomerleau) [reply docs/rules/no-unguarded-async-and-event-operations.md]: Sounds good to me!

File: docs/rules/no-unguarded-async-event-in-ssr.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
+## Disallow Unguarded Async Operations and Event Listeners during SSR (no-unguarded-async-event-in-ssr)
```
Comment (lpomerleau) [inline docs/rules/no-unguarded-async-event-in-ssr.md:1]:
```suggestion
## Disallow Unguarded Async Operations and Event Listeners during SSR (ssr/no-unguarded-async-event)
```

File: README.md [hunk 6]
```diff
@@ -86,6 +86,7 @@ To choose from three configuration settings, install the [`eslint-config-lwc`](h
 | [lwc/no-unsupported-ssr-properties](./docs/rules/no-unsupported-ssr-properties.md)                                   | disallow access of unsupported properties in SSR                   |         |
 | [lwc/no-node-env-in-ssr](./docs/rules/no-node-env-in-ssr.md)                                                         | disallow usage of process.env.NODE_ENV in SSR                      |         |
 | [lwc/valid-graphql-wire-adapter-callback-parameters](./docs/rules/valid-graphql-wire-adapter-callback-parameters.md) | ensure graphql wire adapters are using 'errors' instead of 'error' |         |
+| [lwc/no-unguarded-async-event-in-ssr](./docs/rules/no-unguarded-async-event-in-ssr.md)                               | disallow unguarded async operations and event listeners in ssr     |         |
```
[the remark this answers is not in the document]
Comment (abhagta-sfdc) [reply README.md]: Renamed the rule.Pls review

**Comment by abhagta-sfdc** [comment]
@lpomerleau here is the log for lint scan in es-content-authoring repo with no-unguarded-async-event enabled[
[lint.log](https://github.com/user-attachments/files/17469921/lint.log)
](url)

**Comment by pmdartus** [comment]
Thanks for sharing the ESLint logs @abhagta-sfdc. 

Looking at the results, in it's current shape the `@lwc/lwc/ssr-no-unguarded-async-event` rule creates too many false positive. Here are a couple of examples from the log containing 394 warnings:
- "Unguarded 'await' expression is not allowed during SSR." is emitted 4 times in [`emailUtils.ts`](https://git.soma.salesforce.com/communities/es-content-authoring/blob/5657a44c2d1424d0f319cb10f9121c1df1ed1523/packages/@unified-authoring/components/src/lwc/es_block_builder/emailUtils/emailUtils.ts). There is nothing incompatible with server-side rendering in this file.
- "Unguarded event operation 'dispatchEvent' is not allowed during SSR." is emitted twice in [`event.js`](https://git.soma.salesforce.com/communities/es-content-authoring/blob/5657a44c2d1424d0f319cb10f9121c1df1ed1523/packages/@unified-authoring/components/src/lwc/es_block_builder/utils/event.js). Same observation as above, there is nothing in this file that is indicating that those events are dispatched during server-side rendering.

I wouldn't recommend engineers to adopt this rule. Requiring to wrap all async code and event dispatching in `import.meta.env.SSR` checks will result in worse quality and less maintainable code. This is the type of issue, I expect to be addressed through comprehensive testing unit and integration testing, not through static analysis.

**Comment by abhagta-sfdc** [comment]
@pmdartus
Thank you for the review! As per the description in the work item assigned to us, we were supposed to flag all async/await operations. Your clarification is helpful and much appreciated.
@lpomerleau
Just checking to confirm that we can close this PR(as PM's suggestion is to deal this via omprehensive testing unit and integration testing, not through static analysis)

**Comment by pmdartus** [comment]
Please don't close the PR for now @abhagta-sfdc. This is my take on those rules, but I am not the decision maker here.

**Comment by lpomerleau** [comment]
I'm on board to close this PR. We have [public](https://developer.salesforce.com/docs/platform/lwr/guide/lwr-configure-component-ssr.html?q=await#2-refactor-asynchronous-code) and [internal](https://github.com/salesforce-experience-platform-emu/lwr/blob/main/packages/%40lwrjs/lwc-ssr/best_practices.md#data) doc which discourages the use of async operations. Additionally, async code does not break SSR; it just doesn't get fully executed, so it is still technically portable.

**Comment by abhagta-sfdc** [comment]
Closing this a

**Pull request closed without merging**